lib/basedriver/capabilities.js

```diff
@@ -106,40 +109,55 @@ function stripAppiumPrefixes (caps) {
 
 // Parse capabilities (based on https://www.w3.org/TR/webdriver/#processing-capabilities)
 function parseCaps (caps, constraints = {}, shouldValidateCaps = true) {
+  // If capabilities request is not an object, return error (#1.1)
   if (!_.isObject(caps)) {
     throw new Error('The capabilities argument was not valid for the following reason(s): "capabilities" must be a JSON object.');
   }
 
-  let {alwaysMatch: requiredCaps, firstMatch: allFirstMatchCaps = []} = caps;
-
-  if (!_.isObject(requiredCaps)) {
-    requiredCaps = {};
-  }
+  // Let 'requiredCaps' be property named 'alwaysMatch' from capabilities request (#2) and 'allFirstMatchCaps' be property named 'firstMatch from capabilities request (#3)
+  let {
+    alwaysMatch: requiredCaps = {}, // If 'requiredCaps' is undefined, set it to an empty JSON object (#2.1)
+    firstMatch: allFirstMatchCaps = [{}], // If 'firstMatch' is undefined set it to a singleton list with one empty object (#3.1)
+  } = caps;
 
-  // Reject 'firstMatch' argument if it's not an array (see spec #3.2)
+  // Reject 'firstMatch' argument if it's not an array (#3.2)
   if (!_.isArray(allFirstMatchCaps)) {
     throw new Error('The capabilities.firstMatch argument was not valid for the following reason(s): "capabilities.firstMatch" must be a JSON array or undefined');
   }
 
-  // Strip out the 'appium:' prefix
+  // Strip out the 'appium:' prefix from all
   stripAppiumPrefixes(requiredCaps);
   for (let firstMatchCaps of allFirstMatchCaps) {
     stripAppiumPrefixes(firstMatchCaps);
   }
 
+  // Validate the requiredCaps. But don't validate 'presence' because if that constraint fails on 'alwaysMatch' it could still pass on one of the 'firstMatch' keys
   if (shouldValidateCaps) {
-    requiredCaps = validateCaps(requiredCaps, constraints);
+    requiredCaps = validateCaps(requiredCaps, constraints, {skipPresenceConstraint: true});
   }
 
-  // Validate all of the first match capabilities (see spec #5)
+
+  // Remove the 'presence' constraint for any keys that are already present in 'requiredCaps'
+  // since we know that this constraint has already passed
+  let constraintsCopy = _.clone(constraints);
```

Comment:
I'd rather use .filter on it ;)

Reply:
That doesn't actually work. I tried doing it the 'lodash' way but the problem is it needs to delete just one property of the constraint object and not the whole thing. So something like this....

```javascript
{
  foo: {
    presence: true,
    required: true,
  }
}
```

Needs to be translated to this (assuming that 'foo' presence isn't required anymore)
```javascript
{
  foo: {
    required: true
  }
}
```

Reply:
How about this?

```js
const constraintsCopy = _.toPairs(constraints)
    .filter((x) => !_.keys(requiredCaps).includes(x[0]))
    .reduce((acc, value) => {
       acc[value[0]] = value[1];
       return acc;
    }, {});
```

Reply:
also I assume the name of `constraintsCopy` variable itself might be more specific

Reply:
Yeah I'll rename it `filterredConstraints`. Although I generally agree that I like using functional programming patterns over iterative, I think this may be a case where the iterative way is more clear.